library/core/src/option.rs

```diff
@@ -142,6 +142,9 @@
 //! from `Some::<T>(_)` to `T` (but transmuting `None::<T>` to `T`
 //! is undefined behaviour).
 //!
+//! Finally, it is guaranteed that, for the cases above, `transmute::<_, Option<T>>([0u8; size_of::<T>()])`
+//! is sound and produces `Option::<T>::None`.
```

Comment:
For some of the types, this is actually documented on the inner types instead, like https://doc.rust-lang.org/nightly/std/num/struct.NonZeroU32.html#layout-1

This seems like an obvious yes for `NonNull` and `&` and `&mut`, but I don't know that writing the guarantee *here* is the right way.

Reply:
`NonNull` and `NonZeroX` types are an obvious yes by nature of "there's one value that we have to represent and only one bit pattern for it".  I think that adding the documentation for those is somewhat recent.

While I think that it's a valid concern to avoid documenting things multiple times, I think that it's important that there's *something* here. This is *the* documentation about the NPO generally.

Reply:
We have an accepted RFC that this optimization applies not only to `Option` but also some cases of `Result`. So medium-term `Option` isn't really the right place to put this. We'll have to find some central place describing the idea of null-pointer opt, and then each individual involved type should reference that central place.

Reply:
I'm afraid this might make it sound like we guarantee None is always all zeros, even for `Option<char>` or `Option<ascii::Char>`, which are actually nonzero.

Reply:
Can we guarantee that `None::<&T>` and `None::<NonNull>` are always represented as all zeros? Could we ever support platforms where NULL isn't 0?

Reply:
> We have an accepted RFC that this optimization applies not only to `Option` but also some cases of `Result`. So medium-term `Option` isn't really the right place to put this. We'll have to find some central place describing the idea of null-pointer opt, and then each individual involved type should reference that central place.

Is there a harm in putting this here until the `Result` optimization is implemented, at which point we can move the docs to a central location?

> I'm afraid this might make it sound like we guarantee None is always all zeros, even for `Option<char>` or `Option<ascii::Char>`, which are actually nonzero.

Would it help if we explicitly called out that this guarantee does not apply to all types, and maybe name these as examples?

> Can we guarantee that `None::<&T>` and `None::<NonNull>` are always represented as all zeros? Could we ever support platforms where NULL isn't 0?

This is a question above my pay grade, but I'd expect ~everything to break if we tried to do this 😛

On a more serious note, if we did this, what would the NPO look like? Would we expect to make guarantees about what value constitutes NULL, or would we just guarantee that *some* value constituted NULL, and so we could make guarantees about the size of `Option<T>`? Basically, I'm wondering if we could still write down the `X` in `transmute::<_, Option<T>>(X)`, or if we just can't make any guarantees at all.